File: mobile_restaurant/models/CategoryModels.ts
import { parseBaseResponse, type BaseResponse } from './AuthModels';

// ==================== CategoryModel ====================
export interface CategoryModel {
  id: string;
  name: string;
  description?: string;
}

export function parseCategoryModel(json: any): CategoryModel {
  return {
    id: json.id,
    name: json.name,
    description: json.description ?? undefined,
  };
}

// ==================== CategoryListResponse ====================
export interface CategoryListResponse extends BaseResponse {
  data: CategoryModel[];
}

export function parseCategoryListResponse(json: any): CategoryListResponse {
  const base = parseBaseResponse(json);
  const raw = json.data;
  const list = Array.isArray(raw) ? raw : [];
  return {
    ...base,
    data: list.map(parseCategoryModel),
  };
}

File: lib/Models/CategoryModels.dart
import 'BaseResponse.dart';

// ==================== CategoryModel ====================
class CategoryModel {
  CategoryModel({
    required this.id,
    required this.name,
    this.description,
  });

  final String id;
  final String name;
  final String? description;

  factory CategoryModel.fromJson(Map<String, dynamic> json) {
    return CategoryModel(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
    );
  }
}

// ==================== CategoryListResponse ====================
class CategoryListResponse extends BaseResponse {
  late List<CategoryModel> data;

  CategoryListResponse({
    required int statusCode,
    required String message,
    List<CategoryModel> data = const [],
  }) : super(statusCode: statusCode, message: message) {
    this.data = data;
  }

  factory CategoryListResponse.fromJson(Map<String, dynamic> json) {
    final raw = json['data'];
    final list = raw is List ? raw : const [];
    return CategoryListResponse(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data: list
          .whereType<Map<String, dynamic>>()
          .map(CategoryModel.fromJson)
          .toList(),
    );
  }
}

